File: TypeScriptCodes/src/communication/CommunicationManager.ts
import WebSocket from 'ws';
import { RealtimeHandler } from './RealtimeHandler';
import { MessageQueue } from './MessageQueue';
import { SecureProtocol } from './SecureProtocol';
import { Message } from './Protocol';
import { ErrorHandler, ErrorSeverity } from '../utils/ErrorHandling';
import { ConfigManager } from '../utils/ConfigManager';

type MessageHandler = (sessionId: string, message: Message) => Promise<void>;

export class CommunicationManager {
    private realtimeHandler: RealtimeHandler;
    private messageQueue: MessageQueue;
    private secureProtocol: SecureProtocol;
    private messageHandlers: Map<string, MessageHandler>;
    private config: ConfigManager;

    constructor() {
        this.realtimeHandler = new RealtimeHandler();
        this.messageQueue = new MessageQueue();
        this.secureProtocol = new SecureProtocol();
        this.messageHandlers = new Map();
        this.config = ConfigManager.getInstance();
    }

    async start(host: string = 'localhost', port: number = 8765): Promise<WebSocket.Server> {
        try {
            // Start realtime server
            const server = await this.realtimeHandler.startServer(host, port);

            // Start message queue
            await this.messageQueue.start();

            // Register default handlers
            this.registerDefaultHandlers();

            return server;
        } catch (error) {
            throw ErrorHandler.createError(
                `Failed to start communication manager: ${error}`,
                ErrorSeverity.ERROR,
                'COMMUNICATION_MANAGER'
            );
        }
    }

    async stop(): Promise<void> {
        try {
            await this.messageQueue.stop();
            this.realtimeHandler.stop();
        } catch (error) {
            throw ErrorHandler.createError(
                `Failed to stop communication manager: ${error}`,
                ErrorSeverity.ERROR,
                'COMMUNICATION_MANAGER'
            );
        }
    }

    private registerDefaultHandlers(): void {
        this.realtimeHandler.registerHandler('message', this.handleRealtimeMessage.bind(this));
        this.realtimeHandler.registerHandler('queue', this.handleQueueMessage.bind(this));
    }

    private async handleRealtimeMessage(sessionId: string, message: Message): Promise<void> {
        try {
            if (message.type && this.messageHandlers.has(message.type)) {
                await this.messageHandlers.get(message.type)!(sessionId, message);
            }
        } catch (error) {
            throw ErrorHandler.createError(
                `Realtime message handling failed: ${error}`,
                ErrorSeverity.ERROR,
                'COMMUNICATION_MANAGER'
            );
        }
    }

    private async handleQueueMessage(sessionId: string, message: Message): Promise<void> {
        try {
            await this.messageQueue.enqueue(sessionId, message);
        } catch (error) {
            throw ErrorHandler.createError(
                `Queue message handling failed: ${error}`,
                ErrorSeverity.ERROR,
                'COMMUNICATION_MANAGER'
            );
        }
    }

    registerHandler(messageType: string, handler: MessageHandler): void {
        this.messageHandlers.set(messageType, handler);
    }

    async sendMessage(sessionId: string, message: Message, priority: number = 0): Promise<void> {
        try {
            const connections = this.realtimeHandler['connections'] as Map<string, WebSocket>;
            if (connections.has(sessionId)) {
                await this.realtimeHandler.broadcast(message);
            } else {
                await this.messageQueue.enqueue(sessionId, message, priority);
            }
        } catch (error) {
            throw ErrorHandler.createError(
                `Message sending failed: ${error}`,
                ErrorSeverity.ERROR,
                'COMMUNICATION_MANAGER'
            );
        }
    }

    async processQueuedMessages(): Promise<void> {
        try {
            const queues = this.messageQueue['queues'] as Map<string, any[]>;
            for (const sessionId of queues.keys()) {
                let message: Message | null;
                while ((message = await this.messageQueue.dequeue(sessionId)) !== null) {
                    if (message.type && this.messageHandlers.has(message.type)) {
                        await this.messageHandlers.get(message.type)!(sessionId, message);
                    }
                }
            }
        } catch (error) {
            throw ErrorHandler.createError(
                `Queued message processing failed: ${error}`,
                ErrorSeverity.ERROR,
                'COMMUNICATION_MANAGER'
            );
        }
    }

    getConnectionStats(): Record<string, any> {
        return {
            activeConnections: (this.realtimeHandler['connections'] as Map<string, WebSocket>).size,
            queuedMessages: this.messageQueue.getTotalMessages(),
            messageTypes: Array.from(this.messageHandlers.keys())
        };
    }
} 

File: PythonCodes/src/communication/communication_manager.py
import asyncio
from typing import Dict, Any, Optional, Callable
from .realtime_handler import RealtimeHandler
from .message_queue import MessageQueue
from .secure_protocol import SecureProtocol
from ..utils.error_handling import KAPESITError, ErrorSeverity
from ..utils.config_manager import config_manager

class CommunicationManager:
    def __init__(self):
        self.realtime_handler = RealtimeHandler()
        self.message_queue = MessageQueue()
        self.secure_protocol = SecureProtocol()
        self.message_handlers: Dict[str, Callable] = {}
        
    async def start(self, host: str = "localhost", port: int = 8765):
        try:
            # Start realtime server
            server = await self.realtime_handler.start_server(host, port)
            
            # Start message queue
            await self.message_queue.start()
            
            # Register default handlers
            self._register_default_handlers()
            
            return server
            
        except Exception as e:
            raise KAPESITError(
                f"Failed to start communication manager: {str(e)}",
                severity=ErrorSeverity.ERROR,
                component="COMMUNICATION_MANAGER"
            )
    
    async def stop(self):
        try:
            await self.message_queue.stop()
        except Exception as e:
            raise KAPESITError(
                f"Failed to stop communication manager: {str(e)}",
                severity=ErrorSeverity.ERROR,
                component="COMMUNICATION_MANAGER"
            )
    
    def _register_default_handlers(self):
        self.realtime_handler.register_handler("message", self._handle_realtime_message)
        self.realtime_handler.register_handler("queue", self._handle_queue_message)
    
    async def _handle_realtime_message(self, session_id: str, message: Dict[str, Any]):
        try:
            if "type" in message and message["type"] in self.message_handlers:
                await self.message_handlers[message["type"]](session_id, message)
        except Exception as e:
            raise KAPESITError(
                f"Realtime message handling failed: {str(e)}",
                severity=ErrorSeverity.ERROR,
                component="COMMUNICATION_MANAGER"
            )
    
    async def _handle_queue_message(self, session_id: str, message: Dict[str, Any]):
        try:
            await self.message_queue.enqueue(session_id, message)
        except Exception as e:
            raise KAPESITError(
                f"Queue message handling failed: {str(e)}",
                severity=ErrorSeverity.ERROR,
                component="COMMUNICATION_MANAGER"
            )
    
    def register_handler(self, message_type: str, handler: Callable):
        self.message_handlers[message_type] = handler
    
    async def send_message(self, session_id: str, message: Dict[str, Any], priority: int = 0):
        try:
            if session_id in self.realtime_handler.connections:
                await self.realtime_handler.broadcast(message)
            else:
                await self.message_queue.enqueue(session_id, message, priority)
        except Exception as e:
            raise KAPESITError(
                f"Message sending failed: {str(e)}",
                severity=ErrorSeverity.ERROR,
                component="COMMUNICATION_MANAGER"
            )
    
    async def process_queued_messages(self):
        try:
            for session_id in list(self.message_queue.queues.keys()):
                while message := await self.message_queue.dequeue(session_id):
                    if "type" in message and message["type"] in self.message_handlers:
                        await self.message_handlers[message["type"]](session_id, message)
        except Exception as e:
            raise KAPESITError(
                f"Queued message processing failed: {str(e)}",
                severity=ErrorSeverity.ERROR,
                component="COMMUNICATION_MANAGER"
            )
    
    def get_connection_stats(self) -> Dict[str, Any]:
        return {
            "active_connections": len(self.realtime_handler.connections),
            "queued_messages": self.message_queue.get_total_messages(),
            "message_types": list(self.message_handlers.keys())
        } 